stm32/mboot: Add support for signed and encrypted DFU files.

andrewleech opened the pull request:
This PR adds optional support for signed and encrypted dfu images in stm32 mboot.

This requires additional settings in the board config and requires `pyhy` python module to be installed into `python3` used when building mboot, eg:
```
    pip3 install pyhy
```
In addition to the standard changes made to mpconfigboard.mk to enable mboot, for encrypted support you also need to add:
```
    MBOOT_ENCRYPTED = 1
```
Also, alongside the existing `TEXT0_ADDR` setting the length of application 
flash space is needed, so for an F767 with 2MB of flash where the application 
starts after the 32K of mboot space, it will equal `2048K - 32K == 0x1F8000`:
```
    TEXT0_ADDR = 0x08008000
    TEXT0_LENGTH = 0x1F8000
```
You will also need to generate a compatible pair of public and private key files, which should be stored in the board definition folder (next to mpconfigboard.mk):
```
    python3 micropython/tools/mboot_generate_keys.py micropython/ports/stm32/boards/PYBD_SF6
    cd micropython/ports/stm32
    make BOARD=PYBD_SF6
```
Once you build the firmware, the `firmware.dfu` file will contain the encrypted and signed binaries.

There will also be an extra `firmware.ftr.bin` and `firmware.ftr.dfu` pair of files which contain an
un-encrypted copy of the footer with the encryption key, either of which can be flashed along with mboot to a fresh device to prepare it for decrypting the full `firmware.dfu`.

---

### Implementation details

The PR utilisies the encryption library [libhydrogen](https://github.com/jedisct1/libhydrogen) which provides both the asymetric signing and symetric encryption used in this PR in a high level fashion suitable for non-cryptography specialists to use safely. It's small enough to fit in the existing mboot flash space and uses modern secure ciphers.

This system works without needing any support from the application and without flash space for a second copy of the firmware. 

The encryption script included (`tools/mboot_signed_dfu.py`) breaks the firmware into chunks (16K by default) each of which is individually zipped, signed and encrypted. 
Each chunk is included in the dfu fiile and sent to the device using standard dfu tools, where it's first stored in ram, before its signature is checked for validation.
If that passes, the chunk is then decrypted. The decryption also included a hash behind the scenes to ensure it's valid.
If that passes, the chunk is un-zipped.
If that passes, the sector of flash is erased (if it hasn't already been erased by a prior chunk), then the chunk is written to flash.

Once all firmware chunks are transmitted a hash check is triggered. This confirms the sha256 of the entire firmware image as stored in flash. If that passes the validity bits at the start of the application flash are set. These are read my mboot at startup before jumping to the application.

When encryption is enabled, read support is disabled in mboot. 
The erase command in dfu is also ignored as the flash sectors are tracked and erased as needed only after a transmitted chunk has been validated. This does mean that chunks need to be sent in order, however standard dfu tools all seem to do this by default.

As mentioned in the previous section, there is also a footer included with the firmware. 
This footer contains the asymmetric public key, used to validate the signature and also used as the symmetric key for decryption.
The footer also contains a version number string (not explicitly used currently) and details about each fw image included (normally two sections for builds with internal flash storage, one otherwise) including its address, length and sha256 hash. 

Another potential future extension is to use the version number in the footer to prevent downgrades if desitred. This feature has been been investigated yet however.

---

### firmware footer location

It's currently stored in the final 96 bytes (0x60 hex) of flash on the chip. 
I initially planned to store it either in the final space of the mboot sector, or the start of the application flash space, but have ended up with this implementation. This is very much open to discussion however.

Storing it in mboot space meant it could not be updated with the application, thereby invalidating any of the other details such as version/length/crc etc. It would only be of use for the encrpytion key, but again, make that difficult to update in future if needed.

Storing the details at the start of application space required moving the start of the application forward by 512 bytes (0x200) as this appears to be a requirement for the interrupt ventor table system in the stm range, it needs to be aligned to blocks of this size in most if not all chips. This means you lose 416 bytes of flash, not insignificant on some parts.

I have also thought about storing these details inside the application space, just after the interrupt vector. This would mean I'd need to consider the vector table as one separate section of firmware, then slice in the header/footer data, then continue with the rest of the firmware. This may well be the best option and I'm still considering implementing it, but wanted to get the current working code up for review/discussion first.

---

Based on discussions in https://github.com/micropython/micropython/issues/5267

dpgeorge commented:
Thanks for this, a great feature to have!

It's probably worth discussing encryption vs signing.  The PR here does encryption only which means someone could still modify the encrypted firmware to have random data in it (leaving the first section unchanged so it still had a valid IRQ vector table) and it would then crash when executed.  If it was signed then this wouldn't be possible.

Also, it looks like the firmware is encrypted before compression, meaning compression is no longer useful.  Did you consider encrypting after compression?  This would also somewhat prevent random modification of the data because in most cases random modification would invalidate the compressed stream (and so tampering would be caught by the decompression stage).

andrewleech commented:
You're correct, I haven't looked at signing at all, just encryption. Considering our requirement is to prevent people flashing/running invalid code this really isn't enough, I didn't think about what happens when you just send random data. 
To prevent invalid code being run it definitely needs to be signed or validated in some similar way.

I didn't think the live flashing with dfu over usb uses compression at all? 
The dfu util on computer unpacks the dfu format then sends the raw data to the board to be flashed.
If the dfu file is accidentally corrupted the crc fail on it should be picked up by dfu util on pc before being sent to the board, but that does nothing to stop someone intentiaonally causing trouble.

My understanding is compression is only used for dfu files that are copied to the unit which then triggers the mboot to install it from the application code. My goal is to still allow live flashing from mboot usb dfu in case the main application has a bug that breaks the update orchestrating and therefore bricks the unit.

andrewleech commented:
Looking some more, I should probably ditch AES and use http://kmackay.ca/micro-ecc/
It look looks it could provide asymmetric ECDSA in only 2-3KB of flash and be used to create a secure signature system and maybe the encryption as well (maybe with separate keys the other way around it something). Certainly warrants since more research.

hoihu commented:
> It's probably worth discussing encryption vs signing.

I have not expressed myself clearly about that in my comments in #5267 sorry. Our application  definetly needs signing over encryption. That's where the idea came up of leaving mboot moreless unchanged and just do the heavylifting of asymetric crypto verification in the uPy application. 

It's also possible to sign the encrypted content of the image I believe. Decrypting could then be done in the uPy application also - after the signature has been verified. Or using a bootloader which can handle the decryption - which this PR provides.

I would strongly vote to use a well-tested crypto library (e.g. mbedTLS has also functions to check signatures). That's where I have mixed feeling if the bootloader itself has to check and decrypt stuff.

andrewleech commented:
In an effort to minimise the amount of [rolling my own crypto](https://www.vice.com/en_us/article/wnx8nq/why-you-dont-roll-your-own-crypto) involved, [libhydrogen](https://github.com/jedisct1/libhydrogen) is probably a much better choice.
Written by the author of libsodium (a very highly respected crypto lib), libhydrogen is a modern cryto library targeting constrained environments.

> The Hydrogen library is a small, easy-to-use, hard-to-misuse cryptographic library.
> Instead of low-level primitives, it exposes simple functions to solve common problems that cryptography can solve.

Basically, amongst other things, it provides both Secret-key encryption and Public-key signatures. 

I havne't tried compiling it in for a size check yet, but the comparisons in [Secure Firmware Updates for Constrained IoT Devices Using Open Standards](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=8725488) look promising:
Cortex m4:
`Libhydrogen signature verification |   time:  .24s   | Stack .44kb | Binary  2.1kb |`
Note: that link is a really good reference article on the topic, though I haven't finished reading it yet

andrewleech commented:
Ok, I'm working on this again this week.

Asymmetric signing of the firmware seems like a guaranteed best practice approach, we (the firmware owner) keep the private key safe, devices in the field only have public. 
Hash the (encrpyted) firmware blob, sign the hash, transmit it along with blob, firmware checks its locally generated hash matches the decrypted hash. Identity confirmed.

Encryption I'm less certain about. I initially assumed asymmetric would be better here too, but if my general understanding is correct this isn't the case. 

For encryption of the fw blob to work I believe it needs the private key on the device itself, with the public key being used to do the actual encrypting, eg. https://github.com/jedisct1/libhydrogen/wiki/N-variant

As such if the private key already needs to be kept secure inside the device there's no real benefit to it being asymmetric, we might as well use the same secret on the server for symmetric encryption?

The security of the symetric encryption can be somewhat increased by including a (new) symetric key with each firmware package, also encrypted with the same public/private key as the signature, so there's just the one public key needing to be stored in the device. This should be treated as the secret symetric key though really, as it's used to retrieve it.

As the firmware isn't sent over a live network there's no possibility for device and server to do any kind of handshaking for higher security mechanisms.

Does anyone know of an encryption scheme that makes more sense than this?

Continuing with my goal of having this all work over USB DFU without needing staging flash, I'm thinking of breaking the firmware down into these chunks which are small enough to fit in the target cpu's ram.
Each of these chunks can then be encrypted and signed and stored as separate 'elements' in the dfu file, this way each element can have the appropriate start address even if the length of the (encrypted and signed) bytes sent over the wire no longer matches the number of bytes written to flash. All that matters is the chunk length is small enough for the free ram available in the particular target chip.

andrewleech commented:
@dpgeorge @jimmo @hoihu I've just pushed my brand new signed and encrypted mboot, along with a detailed new top post. I'd welcome any feedback you have.

pi-anl committed:
lib/libhydrogen: Add libhydrogen crypto library source.

From https://github.com/jedisct1/libhydrogen/tree/fda46564cee912140ab7ba3ef0665bd5c74fcc06

pi-anl committed:
stm32/mboot: Add support for signed and encrypted DFU files.

andrewleech commented:
@dpgeorge I've just pushed a new version of this branch with significant changes.

The first 6 commits should each be looked at separately, they contain features not immediately related to the encryption feature. I'm happy to split them out into separate PR's if that's easier.

The final two contain all the encryption work. 
The top post has also been updated to reflect the current functionality.

andrewleech commented:
I've now split out most of the underlying changes into separate PR's, the mass erase one in particular got larger than initially anticipated.

As such only the final two commits in this branch are related to the signing/encrypting support. The rest of the commits are covered in:
* stm32/mboot: Support mass erase dfu command #5730
* stm32/mboot & pydfu USB vid and pid #5725
* tools/pydfu: Add args for vid, pid & exit with cleaner error handling. #5724
* stm32/mboot: Error messages and invalid addresses #5728

mattytrentini commented on ports/stm32/mboot/README.md:
```diff
@@ -137,6 +137,42 @@ The firmware to load must be a gzip'd DfuSe file (.dfu.gz).
+
+    MBOOT_ENCRYPTED = 1
+
+Also, alongside the existing `TEXT0_ADDR` setting the length of application 
+flash space is needed, so for an F767 with 2MB of flash where the application 
+starts after the 32K of mboot space, it will equal `2048K - 32K == 0x1F8000`:
+
+    TEXT0_ADDR = 0x08008000
+    TEXT0_LENGTH = 0x1F8000
+
```
Should add a comma:

```suggestion
Also, alongside the existing `TEXT0_ADDR` setting, the length of application 
```

Also, it would be helpful to include _why_ `TEXT0_LENGTH` is necessary. Something along the lines of: "`TEXT0_LENGTH' is required to indicate where to store the checksum required to validate the encrypted binary".

mattytrentini commented on ports/stm32/mboot/main.c:
```diff
@@ -731,9 +952,227 @@ int do_write(uint32_t addr, const uint8_t *src8, size_t len) {
+                dfu_state.error = MBOOT_ERROR_STR_INVALID_SIG_IDX;
+                return -1;
+            } else {
+                return 0;
```
Is this a legitimate error condition? Should we assert?

andrewleech replied:
There were some normal use cases where it gets to here, I think it was related to not having the full amount of data yet (at start or end of flash/chunk) but I can't remember exactly what it was now. The not having a full chunk should in theory be picked up by the final if just above but I think there was something that it didn't catch.

mattytrentini commented on ports/stm32/mboot/main.c:
```diff
@@ -731,9 +952,227 @@ int do_write(uint32_t addr, const uint8_t *src8, size_t len) {
+    uint8_t ret = 0;
+    if (gzip) {
+        TINF_DATA tinf = {0};
+        uint8_t buf[512*8];
```
Must these numbers be fixed? What defines the size of the buffer? Best to describe the constants with variable names.

mattytrentini commented on ports/stm32/mboot/main.c:
```diff
@@ -905,61 +1346,72 @@ uint8_t i2c_slave_process_tx_byte(void) {
+static size_t get_timeout_ms() {
+    if (dfu_state.wBlockNum == 0) {
+        // download control commands
+        if (dfu_state.wLength >= 1 && dfu_state.buf[0] == 0x41) {
```
What does `0x41` indicate?

andrewleech replied:
I'm actually not too sure, the conditionals were copied from the dfu_process_dnload function below.

mattytrentini commented on ports/stm32/mboot/main.c:
```diff
@@ -905,61 +1346,72 @@ uint8_t i2c_slave_process_tx_byte(void) {
+                #if MBOOT_ENCRYPTED
+                return 0;
+                #else
+                return 500;
```
500 if not encrypted?

mattytrentini commented on ports/stm32/mboot/main.c:
```diff
@@ -905,61 +1346,72 @@ uint8_t i2c_slave_process_tx_byte(void) {
+                return 500;
+                #endif
+            }
+        } else if (dfu_state.wLength >= 1 && dfu_state.buf[0] == 0x21) {
```
0x21?

mattytrentini commented on ports/stm32/mboot/stm32_generic.ld:
```diff
@@ -5,7 +5,8 @@
 MEMORY
 {
-    FLASH_BL (rx)   : ORIGIN = 0x08000000, LENGTH = 32K /* sector 0 (can be 32K) */
+    /* top 32 bytes of flash used to hold encryption key */
```
'top 32'? Isn't it the last 32? Or maybe I'm thinking about it wrong...

mattytrentini commented:
Looks good Andrew! Only other points apart from the specific inline notes are that `mboot/main.c` has become quite large; might make sense to split it up. And it's not clear to me which processors this would be supported on is it really any MBOOT device?

pi-anl committed:
tools/intelhex: Add flattened copy of intelhex tool
From https://github.com/python-intelhex/intelhex/blob/5c15003d08bf53512d5eaa54c7d2825a8b5c37cc/intelhex/__init__.py
Includes intelhex/compat.py and intelhex/getsizeof.py at end.

andrewleech reviewed (commented):
I've just pushed significant changes addressing most of the review comments from @mattytrentini

andrewleech commented:
I was able to find a define from the STM headers for the length of flash in the CPU being compiled for so was able to refactor out the need to set `TEXT0_LENGTH` at all.

This flowed through into needing the firmware footer and mboot key object linked into the application files which should make it a cleaner / easier to follow implementation.

I've also been able to rebase this changeset to not be sitting on top of my mass erase pr, making this changeset cleaner and smaller, even with a couple of functional change overlaps between the two.

@dpgeorge I tend to agree with @mattytrentini that mboot/main.c could benefit from being split out into some separate files. That split would be better done in its own commit before I go and add all this on top. 
Would you agree on separate c files pulling each of the following groups of code out of main?
* `flash` define/read/erase/write functions
* USB
* i2c
Some of the dfu global structures / util functions might be a bit messy to share between files, though if they're no longer made static and extern declared in the common dfu it should come together without much disruption.

dpgeorge commented:
See a new version of this in #6771

dpgeorge commented:
I've now merged #6771 .  Thank you @andrewleech for all the hard work on this PR, which formed the basis for that work.
